crypto/s2n_hkdf.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <stdio.h>
+
+#include "error/s2n_errno.h"
+
+#include "crypto/s2n_hmac.h"
+
+#include "utils/s2n_blob.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+#define MAX_DIGEST_SIZE 64 /*current highest is SHA512*/
+
+//Reference: RFC 5869
+
+int s2n_hkdf_extract(s2n_hmac_algorithm alg,
+                     const struct s2n_blob *salt,
+                     const struct s2n_blob *key,
+                     struct s2n_blob *pseudo_rand_key)
+{
+    struct s2n_hmac_state hmac;
+    pseudo_rand_key->size = s2n_hmac_digest_size(alg);
+
+    GUARD(s2n_hmac_init(&hmac, alg, salt->data, salt->size));
+    GUARD(s2n_hmac_update(&hmac, key->data, key->size));
+    GUARD(s2n_hmac_digest(&hmac, pseudo_rand_key->data, pseudo_rand_key->size));
+
+    return 0;
+}
+
+int s2n_hkdf_expand(s2n_hmac_algorithm alg,
+                    const struct s2n_blob *pseudo_rand_key,
+                    const struct s2n_blob *info,
+                    struct s2n_blob *out_key)
+{
+    uint8_t prev[MAX_DIGEST_SIZE] = "";
+
+    uint32_t done_len = 0;
+    uint32_t hash_len = s2n_hmac_digest_size(alg);
+    uint32_t total_rounds = out_key->size / hash_len;
+    if(out_key->size % hash_len)
+        total_rounds++;
+
+    if(total_rounds > 255)
```

Comment:
Since 255*hash_digest_len is the max as defined by the RFC, consider using S2N_ERROR here with something like S2N_ERR_HKDF_OUTPUT_SIZE. We can also add a unit test to exercise this error.